File: 1814.count-nice-pairs-in-an-array.py
#
# @lc app=leetcode id=1814 lang=python3
#
# [1814] Count Nice Pairs in an Array
#

# @lc code=start
from collections import Counter
from math import comb
class Solution:
    def countNicePairs(self, nums: list[int]) -> int:
        # def rev(num):
        #     y = 0
        #     while num:
        #         y = y * 10 + num % 10
        #         num //= 10
        #     return y
        # ans = 0
        # hashmap = Counter()
        # for num in nums:
        #     if num - rev(num) in hashmap:
        #         ans += hashmap[num - rev(num)]
        #     hashmap[num - rev(num)] += 1
        # return ans % (10**9 + 7)
                # nums[i] + rev(nums[j]) == nums[j] + rev(nums[i]) can be rewritten as
        # nums[i] - rev(nums[i]) == nums[j] - rev(nums[j])
        
        rev_count = Counter([n - int(str(n)[::-1]) for n in nums]) # freq of each nums[i] - rev(nums[i])
        
        # for each group of same nums[i] - rev(nums[i]), find total pair combinations
        # return all pairs modulo 10^9 + 7
        return sum(comb(count, 2) for count in rev_count.values()) % (10**9 + 7)
        
# @lc code=end

File: 1814.count-nice-pairs-in-an-array.java
/*
 * @lc app=leetcode id=1814 lang=java
 *
 * [1814] Count Nice Pairs in an Array
 */

// @lc code=start
import java.util.HashMap;
import java.util.Map;
class Solution {
    public int countNicePairs(int[] nums) {
        //i < j
        //nums[i] + rev(nums[j]) == nums[j] + rev(nums[i])
        //nums[i] - rev(nums[i]) == nums[j] - rev(nums[j]) Equivalent solves the problem
        double MOD = Math.pow(10, 9) + 7;
        int n = nums.length; 
        int ans = 0;
        Map<Integer, Integer> map = new HashMap<>();
        for(int i = 0; i < n; i++){
            int diff = nums[i] - getReverse(nums[i]);
            ans += map.getOrDefault(diff, 0);
            ans %= MOD;
            map.put(diff, map.getOrDefault(diff, 0) + 1);
        }
        return ans;

    }
    private int getReverse(int num){
        int rev = 0;
        while (num > 0){
            rev = 10 * rev + num % 10;
            num /= 10;
        }
        return rev;
    }
}
// @lc code=end

